File: ios/ExpoNativeMqttModule.swift
import ExpoModulesCore
import CocoaMQTT
import Foundation

struct MqttWillOption {
  let topic: String
  let payloadBase64: String
  let qos: Int
  let retained: Bool
}

struct MqttConnectOptions {
  let clientId: String
  let cleanSession: Bool
  let keepAlive: UInt16
  let allowUntrustedCA: Bool
  let maxReconnectAttempts: Int
  let reconnectDelay: Double
  let autoReconnect: Bool
  let maxPayloadSize: Int
  let will: MqttWillOption?
}

public class ExpoNativeMqttModule: Module {
  private var mqtt: CocoaMQTT?
  private var storedOptions: MqttConnectOptions?
  private var subscribedTopics: [String: Int] = [:] // topic -> qos
  private var reconnectTimer: Timer?
  private var isDisconnectingManually = false
  private var reconnectAttempts = 0
  private let lock = NSLock()

  private var connectPromise: Promise?
  private var disconnectPromise: Promise?
  private var pendingSubscribePromises: [String: Promise] = [:]
  private var pendingUnsubscribePromises: [String: Promise] = [:]
  private var pendingPublishPromises: [UInt16: Promise] = [:]
  
  deinit {
    lock.lock()
    reconnectTimer?.invalidate()
    reconnectTimer = nil
    mqtt?.username = nil
    mqtt?.password = nil
    mqtt = nil
    lock.unlock()
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoNativeMqtt")

    Events(
      "onMqttConnected",
      "onMqttDisconnected",
      "onMqttError",
      "onMqttMessageReceived",
      "onMqttSubscribed",
      "onMqttUnsubscribed",
      "onMqttReconnecting"
    )

    AsyncFunction("connect") { (brokerUrl: String, username: String?, password: String?, options: [String: Any], promise: Promise) in
      guard let url = URL(string: brokerUrl) else {
        promise.reject("INVALID_URL", "Invalid broker URL")
        return
      }

      let host = url.host ?? brokerUrl
      let port = UInt16(url.port ?? 1883)
      
      let willOpt: MqttWillOption?
      if let willDict = options["will"] as? [String: Any],
         let topic = willDict["topic"] as? String,
         let payloadBase64 = willDict["payloadBase64"] as? String {
        willOpt = MqttWillOption(
          topic: topic,
          payloadBase64: payloadBase64,
          qos: willDict["qos"] as? Int ?? 0,
          retained: willDict["retained"] as? Bool ?? false
        )
      } else {
        willOpt = nil
      }

      let parsedOptions = MqttConnectOptions(
        clientId: options["clientId"] as? String ?? "expo-mqtt-\(UUID().uuidString)",
        cleanSession: options["cleanSession"] as? Bool ?? false,
        keepAlive: options["keepAlive"] as? UInt16 ?? 60,
        allowUntrustedCA: options["allowUntrustedCA"] as? Bool ?? false,
        maxReconnectAttempts: options["maxReconnectAttempts"] as? Int ?? 10,
        reconnectDelay: options["reconnectDelay"] as? Double ?? 5000.0,
        autoReconnect: options["autoReconnect"] as? Bool ?? true,
        maxPayloadSize: options["maxPayloadSize"] as? Int ?? 262144,
        will: willOpt
      )

      self.lock.lock()
      self.storedOptions = parsedOptions
      self.isDisconnectingManually = false
      self.reconnectAttempts = 0
      self.reconnectTimer?.invalidate()
      self.reconnectTimer = nil
      self.connectPromise = promise
      self.lock.unlock()

      // Disconnect existing client to prevent resource leaks
      if let existingClient = self.mqtt {
        existingClient.delegate = nil
        existingClient.disconnect()
        self.mqtt = nil
      }

      self.mqtt = CocoaMQTT(clientID: parsedOptions.clientId, host: host, port: port)
      self.mqtt?.username = username
      self.mqtt?.password = password
      self.mqtt?.keepAlive = parsedOptions.keepAlive
      self.mqtt?.cleanSession = parsedOptions.cleanSession
      self.mqtt?.delegate = self
      
      if let will = parsedOptions.will, let willData = Data(base64Encoded: will.payloadBase64) {
        let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(will.qos)) ?? .qos0
        let willMessage = CocoaMQTTMessage(topic: will.topic, payload: [UInt8](willData), qos: cocoaQos, retained: will.retained)
        self.mqtt?.willMessage = willMessage
      }

      // SSL Setup
      if brokerUrl.hasPrefix("ssl://") || brokerUrl.hasPrefix("wss://") {
        self.mqtt?.enableSSL = true
        if parsedOptions.allowUntrustedCA {
          #if !DEBUG
          self.lock.lock()
          let capturedPromise = self.connectPromise
          self.connectPromise = nil
          self.lock.unlock()
          capturedPromise?.reject("SECURITY_ERROR", "allowUntrustedCA is not permitted in release builds due to security risks.")
          return
          #endif
          self.mqtt?.allowUntrustCACertificate = true
        }
        // TODO: Implement certificate pinning in a future build.
        // The code below does certificate pinning (when implemented)
      }

      let success = self.mqtt?.connect() ?? false
      if !success {
        self.lock.lock()
        let capturedPromise = self.connectPromise
        self.connectPromise = nil
        self.lock.unlock()
        capturedPromise?.reject("CONNECT_FAILED", "Failed to initiate connection")
      }
    }

    AsyncFunction("disconnect") { (promise: Promise) in
      self.lock.lock()
      self.isDisconnectingManually = true
      self.reconnectTimer?.invalidate()
      self.reconnectTimer = nil
      self.storedOptions = nil
      self.reconnectAttempts = 0
      self.disconnectPromise = promise
      self.lock.unlock()

      guard let client = self.mqtt else {
        self.lock.lock()
        self.disconnectPromise = nil
        self.lock.unlock()
        promise.resolve("Disconnected")
        return
      }

      client.disconnect()
    }

    AsyncFunction("subscribe") { (topic: String, qos: Int, promise: Promise) in
      self.lock.lock()
      self.subscribedTopics[topic] = qos
      self.pendingSubscribePromises[topic] = promise
      let client = self.mqtt
      self.lock.unlock()
      
      guard let mqttClient = client else {
        promise.reject("NOT_CONNECTED", "No active MQTT connection")
        return
      }

      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos0
      mqttClient.subscribe(topic, qos: cocoaQos)
    }

    AsyncFunction("unsubscribe") { (topic: String, promise: Promise) in
      self.lock.lock()
      self.subscribedTopics.removeValue(forKey: topic)
      self.pendingUnsubscribePromises[topic] = promise
      let client = self.mqtt
      self.lock.unlock()
      
      guard let mqttClient = client else {
        promise.reject("NOT_CONNECTED", "No active MQTT connection")
        return
      }

      mqttClient.unsubscribe(topic)
    }

    AsyncFunction("publish") { (topic: String, base64Payload: String, qos: Int, retained: Bool, promise: Promise) in
      self.lock.lock()
      let client = self.mqtt
      self.lock.unlock()
      
      guard let mqttClient = client else {
        promise.reject("NOT_CONNECTED", "No active MQTT connection")
        return
      }
      
      guard let data = Data(base64Encoded: base64Payload) else {
        promise.reject("INVALID_PAYLOAD", "Invalid base64 payload")
        return
      }
      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos0
      let bytes = [UInt8](data)
      let msgId = mqttClient.publish(topic, withBytes: bytes, qos: cocoaQos, retained: retained)
      
      if qos == 0 || msgId == nil {
        promise.resolve("Published")
      } else {
        self.lock.lock()
        self.pendingPublishPromises[msgId!] = promise
        self.lock.unlock()
      }
    }
  }

  fileprivate func scheduleReconnect() {
    lock.lock()
    guard !isDisconnectingManually else {
      lock.unlock()
      return
    }
    guard let optionsToUse = storedOptions, optionsToUse.autoReconnect else {
      lock.unlock()
      return
    }

    let maxAttempts = optionsToUse.maxReconnectAttempts
    if maxAttempts > 0 && reconnectAttempts >= maxAttempts {
      lock.unlock()
      sendEvent("onMqttError", ["error": "Max reconnect attempts reached (\(maxAttempts))"])
      return
    }

    let baseDelay = optionsToUse.reconnectDelay
    let factor = pow(2.0, Double(min(reconnectAttempts, 6)))
    let cappedDelay = min(baseDelay * factor, 60000.0)
    let jitter = Double.random(in: 0.75...1.25)
    let delaySeconds = (cappedDelay * jitter) / 1000.0

    reconnectAttempts += 1
    let currentAttempt = reconnectAttempts
    lock.unlock()

    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      self.lock.lock()
      self.reconnectTimer?.invalidate()
      self.reconnectTimer = Timer.scheduledTimer(withTimeInterval: delaySeconds, repeats: false) { [weak self] _ in
        guard let self = self else { return }
        self.lock.lock()
        let isManual = self.isDisconnectingManually
        self.lock.unlock()
        
        if !isManual {
          self.sendEvent("onMqttReconnecting", ["status": "reconnecting", "attempt": currentAttempt])
          DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            self.lock.lock()
            let client = self.mqtt
            self.lock.unlock()
            _ = client?.connect()
          }
        }
      }
      self.lock.unlock()
    }
  }

  fileprivate func resubscribeAll() {
    lock.lock()
    let topicsCopy = subscribedTopics
    let client = mqtt
    lock.unlock()

    for (topic, qos) in topicsCopy {
      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos0
      client?.subscribe(topic, qos: cocoaQos)
    }
  }
}

extension ExpoNativeMqttModule: CocoaMQTTDelegate {
  public func mqtt(_ mqtt: CocoaMQTT, didConnectAck ack: CocoaMQTTConnAck) {
    lock.lock()
    let promise = connectPromise
    connectPromise = nil
    reconnectAttempts = 0
    lock.unlock()

    if ack == .accept {
      promise?.resolve("Connected")
      sendEvent("onMqttConnected", ["status": "connected"])
      resubscribeAll()
    } else {
      promise?.reject("CONNECT_REFUSED", "Connection refused: \(ack)")
      sendEvent("onMqttError", ["error": "Connection refused: \(ack)"])
    }
  }

  public func mqttDidDisconnect(_ mqtt: CocoaMQTT, withError error: Error?) {
    lock.lock()
    let promise = disconnectPromise
    disconnectPromise = nil
    let isManual = isDisconnectingManually
    let publishPromises = pendingPublishPromises
    pendingPublishPromises.removeAll()
    let subscribePromises = pendingSubscribePromises
    pendingSubscribePromises.removeAll()
    let unsubscribePromises = pendingUnsubscribePromises
    pendingUnsubscribePromises.removeAll()
    lock.unlock()

    // Reject any pending publish promises that will never be acked
    for (_, p) in publishPromises {
      p.reject("DISCONNECTED", "Disconnected before publish was acknowledged")
    }
    for (_, p) in subscribePromises {
      p.reject("DISCONNECTED", "Disconnected before subscribe was acknowledged")
    }
    for (_, p) in unsubscribePromises {
      p.reject("DISCONNECTED", "Disconnected before unsubscribe was acknowledged")
    }

    promise?.resolve("Disconnected")
    sendEvent("onMqttDisconnected", [
      "error": error?.localizedDescription ?? "clean disconnect"
    ])

    if isManual {
      self.mqtt?.username = nil
      self.mqtt?.password = nil
      self.mqtt = nil
    }

    scheduleReconnect()
  }

  public func mqtt(_ mqtt: CocoaMQTT, didReceiveMessage message: CocoaMQTTMessage, id: UInt16) {
    let payloadBytes = message.payload
    
    lock.lock()
    let maxSize = storedOptions?.maxPayloadSize ?? 262144
    lock.unlock()
    
    if payloadBytes.count > maxSize {
      sendEvent("onMqttError", ["error": "Message dropped: payload size (\(payloadBytes.count) bytes) exceeds maximum (\(maxSize) bytes)"])
      return
    }

    let payloadBase64 = Data(payloadBytes).base64EncodedString()

    sendEvent("onMqttMessageReceived", [
      "topic": message.topic,
      "payloadBase64": payloadBase64,
      "qos": Int(message.qos.rawValue),
      "retained": message.retained
    ])
  }

  public func mqtt(_ mqtt: CocoaMQTT, didSubscribeTopics success: NSDictionary, failed: [String]) {
    let topics = success.allKeys.compactMap { $0 as? String }
    lock.lock()
    for topic in topics {
      pendingSubscribePromises.removeValue(forKey: topic)?.resolve("Subscribed")
    }
    for topic in failed {
      pendingSubscribePromises.removeValue(forKey: topic)?.reject("SUBSCRIBE_FAILED", "Failed to subscribe to \(topic)")
    }
    lock.unlock()

    sendEvent("onMqttSubscribed", ["topics": topics])
  }

  public func mqtt(_ mqtt: CocoaMQTT, didUnsubscribeTopics topics: [String]) {
    lock.lock()
    for topic in topics {
      pendingUnsubscribePromises.removeValue(forKey: topic)?.resolve("Unsubscribed")
    }
    lock.unlock()

    for topic in topics {
      sendEvent("onMqttUnsubscribed", ["topic": topic])
    }
  }

  public func mqtt(_ mqtt: CocoaMQTT, didPublishMessage message: CocoaMQTTMessage, id: UInt16) { }

  public func mqtt(_ mqtt: CocoaMQTT, didPublishAck id: UInt16) {
    lock.lock()
    let promise = pendingPublishPromises.removeValue(forKey: id)
    lock.unlock()
    promise?.resolve("Published")
  }

  public func mqttDidPing(_ mqtt: CocoaMQTT) { }

  public func mqttDidReceivePong(_ mqtt: CocoaMQTT) { }
}

File: android/src/main/java/expo/modules/nativemqtt/ExpoNativeMqttModule.kt
package expo.modules.nativemqtt

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.hivemq.client.mqtt.mqtt3.Mqtt3AsyncClient
import com.hivemq.client.mqtt.mqtt3.Mqtt3Client
import com.hivemq.client.mqtt.datatypes.MqttQos
import com.hivemq.client.mqtt.MqttGlobalPublishFilter
import com.hivemq.client.mqtt.mqtt3.message.connect.connack.Mqtt3ConnAck
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.Arrays
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

data class MqttWillOption(
    val topic: String,
    val payload: ByteArray,
    val qos: Int,
    val retained: Boolean
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as MqttWillOption
        if (topic != other.topic) return false
        if (!payload.contentEquals(other.payload)) return false
        if (qos != other.qos) return false
        if (retained != other.retained) return false
        return true
    }

    override fun hashCode(): Int {
        var result = topic.hashCode()
        result = 31 * result + payload.contentHashCode()
        result = 31 * result + qos
        result = 31 * result + retained.hashCode()
        return result
    }
}

data class MqttConnectOptions(
    val clientId: String,
    val cleanSession: Boolean,
    val keepAlive: Int,
    val allowUntrustedCA: Boolean,
    val maxReconnectAttempts: Int,
    val reconnectDelay: Double,
    val autoReconnect: Boolean,
    val maxPayloadSize: Int,
    val will: MqttWillOption?
)

class ExpoNativeMqttModule : Module() {
    private var client: Mqtt3AsyncClient? = null
    private var storedOptions: MqttConnectOptions? = null
    private val subscribedTopics = ConcurrentHashMap<String, Int>() // topic -> qos
    private var isDisconnectingManually = false
    private var reconnectExecutor: ScheduledExecutorService? = null
    private var storedUsername: String? = null
    private var storedPassword: CharArray? = null
    private var reconnectAttempts = 0
    private var isReplacingClient = false
    private val lock = Any()

    override fun definition() = ModuleDefinition {
        Name("ExpoNativeMqtt")

        Events(
            "onMqttConnected",
            "onMqttDisconnected",
            "onMqttError",
            "onMqttMessageReceived",
            "onMqttSubscribed",
            "onMqttUnsubscribed",
            "onMqttReconnecting"
        )

        OnDestroy {
            synchronized(lock) {
                reconnectExecutor?.shutdownNow()
                reconnectExecutor = null
                storedPassword?.let { Arrays.fill(it, '\u0000') }
                storedPassword = null
                storedUsername = null
                client?.disconnect()
                client = null
            }
        }

        AsyncFunction("connect") { brokerUrl: String, username: String?, password: String?, options: Map<String, Any>, willPayload: ByteArray?, promise: Promise ->
            try {
                // Parse options securely
                val parsedOptions = MqttConnectOptions(
                    clientId = options["clientId"] as? String ?: "expo-mqtt-${UUID.randomUUID()}",
                    cleanSession = options["cleanSession"] as? Boolean ?: false,
                    keepAlive = (options["keepAlive"] as? Number)?.toInt() ?: 60,
                    allowUntrustedCA = options["allowUntrustedCA"] as? Boolean ?: false,
                    maxReconnectAttempts = (options["maxReconnectAttempts"] as? Number)?.toInt() ?: 10,
                    reconnectDelay = (options["reconnectDelay"] as? Number)?.toDouble() ?: 5000.0,
                    autoReconnect = options["autoReconnect"] as? Boolean ?: true,
                    maxPayloadSize = (options["maxPayloadSize"] as? Number)?.toInt() ?: 262144,
                    will = (options["will"] as? Map<String, Any>)?.let { willMap ->
                        MqttWillOption(
                            topic = willMap["topic"] as? String ?: "",
                            payload = willPayload ?: ByteArray(0),
                            qos = (willMap["qos"] as? Number)?.toInt() ?: 0,
                            retained = willMap["retained"] as? Boolean ?: false
                        )
                    }
                )

                synchronized(lock) {
                    storedOptions = parsedOptions
                    reconnectAttempts = 0
                    storedUsername = username
                    storedPassword?.let { Arrays.fill(it, '\u0000') }
                    storedPassword = password?.toCharArray()

                    if (reconnectExecutor == null || reconnectExecutor!!.isShutdown) {
                        reconnectExecutor = Executors.newSingleThreadScheduledExecutor()
                    }

                    // Clean up existing client to prevent resource leaks
                    isReplacingClient = true
                    client?.let { existingClient ->
                        try { 
                            existingClient.disconnect().whenComplete { _, _ ->
                                synchronized(lock) {
                                    isReplacingClient = false
                                }
                            }
                        } catch (_: Exception) {
                            isReplacingClient = false
                        }
                    } ?: run {
                        isReplacingClient = false
                    }
                    client = null
                    isDisconnectingManually = false
                }

                // Parse Broker URL
                var host = brokerUrl
                var port = 1883
                try {
                    val uriString = if (!brokerUrl.contains("://")) "tcp://$brokerUrl" else brokerUrl
                    val uri = java.net.URI(uriString)
                    host = uri.host ?: brokerUrl
                    if (uri.port != -1) {
                        port = uri.port
                    }
                } catch (e: Exception) {
                    val cleanUrl = brokerUrl.replace("tcp://", "").replace("ssl://", "").replace("wss://", "").replace("ws://", "")
                    val parts = cleanUrl.split(":")
                    host = parts.getOrNull(0) ?: brokerUrl
                    port = parts.getOrNull(1)?.toIntOrNull() ?: 1883
                }

                val builder = Mqtt3Client.builder()
                    .identifier(parsedOptions.clientId)
                    .serverHost(host)
                    .serverPort(port)
                    .addDisconnectedListener { context ->
                        val error = context.cause
                        sendEvent("onMqttDisconnected", mapOf("error" to (error?.message ?: "Disconnected")))
                        
                        var shouldReconnect = false
                        synchronized(lock) {
                            shouldReconnect = !isDisconnectingManually && !isReplacingClient
                        }
                        if (shouldReconnect) {
                            scheduleReconnect()
                        }
                    }

                // SSL Setup
                if (brokerUrl.startsWith("ssl://") || brokerUrl.startsWith("wss://")) {
                    if (parsedOptions.allowUntrustedCA) {
                        val isDebug = (appContext.reactContext?.applicationInfo?.flags?.and(android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) ?: 0) != 0
                        if (!isDebug) {
                            Log.w("ExpoNativeMqtt", "allowUntrustedCA is enabled in release build. Use caution.")
                        }
                        
                        val trustAllTmf = object : javax.net.ssl.TrustManagerFactory(
                            object : javax.net.ssl.TrustManagerFactorySpi() {
                                override fun engineInit(ks: java.security.KeyStore?) {}
                                override fun engineInit(spec: javax.net.ssl.ManagerFactoryParameters?) {}
                                override fun engineGetTrustManagers(): Array<javax.net.ssl.TrustManager> = arrayOf(
                                    object : javax.net.ssl.X509TrustManager {
                                        override fun checkClientTrusted(chain: Array<out java.security.cert.X509Certificate>?, authType: String?) {}
                                        override fun checkServerTrusted(chain: Array<out java.security.cert.X509Certificate>?, authType: String?) {}
                                        override fun getAcceptedIssuers(): Array<java.security.cert.X509Certificate> = arrayOf()
                                    }
                                )
                            }, null, "TrustAll"
                        ) {}
                        
                        // TODO: Implement certificate pinning in a future build.
                        // The code below does certificate pinning (when implemented)
                        builder.sslConfig()
                            .trustManagerFactory(trustAllTmf)
                            // We purposefully DO NOT disable hostnameVerifier here anymore for security
                            .applySslConfig()
                    } else {
                        // TODO: Implement certificate pinning in a future build.
                        // The code below does certificate pinning (when implemented)
                        builder.sslWithDefaultConfig()
                    }
                }

                val newClient = builder.buildAsync()
                synchronized(lock) {
                    client = newClient
                }

                // Global Publish listener
                newClient.publishes(MqttGlobalPublishFilter.ALL) { publish ->
                    val payloadBytes = publish.payloadAsBytes
                    
                    var maxSize = 262144
                    synchronized(lock) {
                        maxSize = storedOptions?.maxPayloadSize ?: 262144
                    }
                    
                    if (payloadBytes.size > maxSize) {
                        sendEvent("onMqttError", mapOf("error" to "Message dropped: payload size (${payloadBytes.size} bytes) exceeds maximum (${maxSize} bytes)"))
                        return@publishes
                    }

                    sendEvent("onMqttMessageReceived", mapOf(
                        "topic" to publish.topic.toString(),
                        "payload" to payloadBytes,
                        "qos" to publish.qos.code,
                        "retained" to publish.isRetain
                    ))
                }

                // Connect builder
                val connectBuilder = newClient.connectWith()
                    .cleanSession(parsedOptions.cleanSession)
                    .keepAlive(parsedOptions.keepAlive)

                synchronized(lock) {
                    if (!storedUsername.isNullOrEmpty() && storedPassword != null) {
                        connectBuilder.simpleAuth()
                            .username(storedUsername!!)
                            .password(String(storedPassword!!).toByteArray())
                            .applySimpleAuth()
                    }
                }

                parsedOptions.will?.let { willOpt ->
                    try {
                        val willQos = MqttQos.fromCode(willOpt.qos) ?: MqttQos.AT_MOST_ONCE
                        connectBuilder.willPublish()
                            .topic(willOpt.topic)
                            .payload(willOpt.payload)
                            .qos(willQos)
                            .retain(willOpt.retained)
                            .applyWillPublish()
                    } catch (e: Exception) {
                        promise.reject("CONNECT_ERROR", "Failed to configure will message: ${e.message}", e)
                        return@AsyncFunction
                    }
                }

                connectBuilder.send().whenComplete { ack: Mqtt3ConnAck?, throwable: Throwable? ->
                    if (throwable != null) {
                        promise.reject("CONNECT_FAILED", throwable.message, throwable)
                        sendEvent("onMqttError", mapOf("error" to throwable.message))
                    } else {
                        synchronized(lock) {
                            reconnectAttempts = 0
                        }
                        promise.resolve("Connected")
                        sendEvent("onMqttConnected", mapOf("status" to "connected"))
                        resubscribeAll()
                    }
                }
            } catch (e: Exception) {
                promise.reject("CONNECT_ERROR", e.message, e)
            }
        }

        AsyncFunction("disconnect") { promise: Promise ->
            val c: Mqtt3AsyncClient?
            synchronized(lock) {
                isDisconnectingManually = true
                reconnectExecutor?.shutdownNow()
                reconnectExecutor = null
                storedUsername = null
                storedPassword?.let { Arrays.fill(it, '\u0000') }
                storedPassword = null
                storedOptions = null
                reconnectAttempts = 0
                c = client
            }
            
            c?.disconnect()?.whenComplete { _, _ ->
                synchronized(lock) {
                    if (client === c) {
                        client = null
                    }
                }
                promise.resolve("Disconnected")
            } ?: promise.resolve("No active client")
        }

        AsyncFunction("subscribe") { topic: String, qos: Int, promise: Promise ->
            val c: Mqtt3AsyncClient?
            synchronized(lock) {
                subscribedTopics[topic] = qos
                c = client
            }
            
            c ?: run {
                promise.reject("NOT_CONNECTED", "No active MQTT connection", null)
                return@AsyncFunction
            }

            val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE
            
            c.subscribeWith()
                ?.topicFilter(topic)
                ?.qos(mqttQos)
                ?.send()
                ?.whenComplete { _, throwable ->
                    if (throwable != null) {
                        promise.reject("SUBSCRIBE_FAILED", throwable.message, throwable)
                    } else {
                        promise.resolve("Subscribed")
                        sendEvent("onMqttSubscribed", mapOf("topics" to listOf(topic)))
                    }
                }
        }

        AsyncFunction("unsubscribe") { topic: String, promise: Promise ->
            val c: Mqtt3AsyncClient?
            synchronized(lock) {
                subscribedTopics.remove(topic)
                c = client
            }
            
            c ?: run {
                promise.reject("NOT_CONNECTED", "No active MQTT connection", null)
                return@AsyncFunction
            }
            
            c.unsubscribeWith()
                ?.topicFilter(topic)
                ?.send()
                ?.whenComplete { _, throwable ->
                    if (throwable != null) {
                        promise.reject("UNSUBSCRIBE_FAILED", throwable.message, throwable)
                    } else {
                        promise.resolve("Unsubscribed")
                        sendEvent("onMqttUnsubscribed", mapOf("topic" to topic))
                    }
                }
        }

        AsyncFunction("publish") { topic: String, payload: ByteArray, qos: Int, retained: Boolean, promise: Promise ->
            val c: Mqtt3AsyncClient?
            synchronized(lock) {
                c = client
            }
            
            c ?: run {
                promise.reject("NOT_CONNECTED", "No active MQTT connection", null)
                return@AsyncFunction
            }
            
            try {
                val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE

                c.publishWith()
                    ?.topic(topic)
                    ?.payload(payload)
                    ?.qos(mqttQos)
                    ?.retain(retained)
                    ?.send()
                    ?.whenComplete { _, throwable ->
                        if (throwable != null) {
                            promise.reject("PUBLISH_FAILED", throwable.message, throwable)
                        } else {
                            promise.resolve("Published")
                        }
                    }
            } catch (e: Exception) {
                promise.reject("PUBLISH_ERROR", e.message, e)
            }
        }
    }

    private fun scheduleReconnect() {
        var optionsToUse: MqttConnectOptions?
        var isManual: Boolean
        synchronized(lock) {
            optionsToUse = storedOptions
            isManual = isDisconnectingManually
        }
        
        val autoReconnect = optionsToUse?.autoReconnect ?: true
        if (!autoReconnect || isManual) return

        val maxAttempts = optionsToUse?.maxReconnectAttempts ?: 10
        
        var currentAttempt: Int
        synchronized(lock) {
            if (maxAttempts > 0 && reconnectAttempts >= maxAttempts) {
                sendEvent("onMqttError", mapOf("error" to "Max reconnect attempts reached ($maxAttempts)"))
                return
            }
            reconnectAttempts++
            currentAttempt = reconnectAttempts
        }

        val baseDelay = optionsToUse?.reconnectDelay ?: 5000.0
        val attemptFactor = 2.0.pow(min(currentAttempt.toDouble(), 6.0))
        val calculatedDelay = min(baseDelay * attemptFactor, 60000.0)
        val jitter = Random.nextDouble(0.75, 1.25)
        val delayMs = (calculatedDelay * jitter).toLong()

        synchronized(lock) {
            if (isDisconnectingManually) return
            
            if (reconnectExecutor == null || reconnectExecutor!!.isShutdown) {
                reconnectExecutor = Executors.newSingleThreadScheduledExecutor()
            }
            
            reconnectExecutor?.schedule({
                var checkManual: Boolean
                var c: Mqtt3AsyncClient?
                synchronized(lock) {
                    checkManual = isDisconnectingManually
                    c = client
                }
                if (checkManual) return@schedule

                c?.let {
                    sendEvent("onMqttReconnecting", mapOf("status" to "reconnecting", "attempt" to currentAttempt))
                    val cleanSession = optionsToUse?.cleanSession ?: false
                    val keepAlive = optionsToUse?.keepAlive ?: 60
                    
                    val connectBuilder = it.connectWith()
                        .cleanSession(cleanSession)
                        .keepAlive(keepAlive)
                    
                    synchronized(lock) {
                        if (!storedUsername.isNullOrEmpty() && storedPassword != null) {
                            connectBuilder.simpleAuth()
                                .username(storedUsername!!)
                                .password(String(storedPassword!!).toByteArray())
                                .applySimpleAuth()
                        }
                    }
                    
                    // Re-apply will message on reconnect
                    optionsToUse?.will?.let { willOpt ->
                        try {
                            val willQos = MqttQos.fromCode(willOpt.qos) ?: MqttQos.AT_MOST_ONCE
                            connectBuilder.willPublish()
                                .topic(willOpt.topic)
                                .payload(willOpt.payload)
                                .qos(willQos)
                                .retain(willOpt.retained)
                                .applyWillPublish()
                        } catch (e: Exception) {
                            // Ignore exception during reconnect will parsing
                        }
                    }
                    
                    connectBuilder.send().whenComplete { _, throwable ->
                        if (throwable == null) {
                            synchronized(lock) {
                                reconnectAttempts = 0
                            }
                            sendEvent("onMqttConnected", mapOf("status" to "connected"))
                            resubscribeAll()
                        }
                    }
                }
            }, delayMs, TimeUnit.MILLISECONDS)
        }
    }

    private fun resubscribeAll() {
        val subscriptions = mutableMapOf<String, Int>()
        val c: Mqtt3AsyncClient?
        synchronized(lock) {
            subscriptions.putAll(subscribedTopics)
            c = client
        }
        
        for ((topic, qos) in subscriptions) {
            val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE
            c?.subscribeWith()
                ?.topicFilter(topic)
                ?.qos(mqttQos)
                ?.send()
        }
    }
}
